proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+resources entering the function, with just one leaving. However, when exposed as
+an Interface Type function, all these resources must be created and properly
+disposed of within the adapter code itself.
+
+This note focuses on the techniques that enable this to be achieved reliably.
+
+## Exporting the Chooser
+
+The Interface Type function signature for our chooser is simple: the function
+takes two `string`s and returns one:
+
+```wasm
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+  ...
+)
+```
+
+One of the specific challenges in this scenario is the handling of shared
+pointers. We 'require' that the core `nondeterministic_choice` function honor
+the semantics of proper reference counting the input arguments and the returned
+string.
+
+However, there are actually _two_ schemes in play in this scenario: the
+Interface Types management of resources and the C++ implementation of
+`shared_ptr`. Effectively, in addition to lifting and lowering the `string`
+values we must also lift and lower the ownership between Interface Types and
+core WASM.
+
+In this scenario we take a somewhat simplified view of C++'s implementation of
+shared pointers: a shared pointer is implemented as a pair consisting of a
+reference count and a raw pointer to the shared resource.
+
+>Note: In practice, the C++ implementation of shared pointers is somewhat more
+>complex; for reasons that are not important here.
+
+This results in _two_ memory allocations for the shared resource: one for the
+resource itself and one for the pointer structure -- which contains a reference
+count and a raw pointer to the resource.
+
+
+```wasm
+(@interface func $stralloc-x
+  (param $str string)
+  (result owned<i32>)        ;; return an owned shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-x"     ;; create string entity of right size
+  own (i32)
+    call-export "free-x"     ;; temporary, until safely re-owned
+  end
+  let (local $ptr owned<i32>) (result owned<i32>)
+    local.get $str
+    local.get $ptr
+	owned.access             ;; access the owned pointer
+	local.get $str
+	string.size
+    utf8.from.string
+	local.get $ptr
+  end
+)
+
+(@interface func $shared-x
+  (param $ptr owned<i32>)
+  (result owned<i32>)        ;; return a shared ptr of the entity
+
+  local.get $ptr
+  owned.release              ;; access pointer and remove ownership
+  i32.const 1                ;; initial reference count of 1
+  call-export "shared-builder-x" ;; make it a shared ptr pair
+  own (i32)                  ;; re-own it, with a different destructor
+    call-export "shared-release-x"
+  end
+end
+)
+
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+
+  local.get $left
+  invoke-func $stralloc-x
+  invoke-func $shared-x     ;; make it a shared ptr
+  let (local $lpo owned<i32>) (result string)
```

Comment:
minor nit: current function-references proposal requires this the other way around:
```
let (result string) (local $lpo owned<i32>)
```

TBH, I think it's more clear the way you have it, but I thought I'd mention.

Reply:
Agreed it reads more naturally to put results second, making `let` more like a function signature where `param`s have been replaced by `local`s.  Maybe we should change the function-references proposal...

Reply:
Ah, but there's a subtlety to `let` that I had missed. `let` introduces a new block, so it has a blocktype, which means w/ the multi-value proposal it's possible to have params and locals:

```wasm
f32.const 1
i32.const 2
let (param i32) (result i64) (local f32)
  ;; i32:2 is still on top of the stack
  ;; f32:1 was consumed from the stack, and is now local 0
  ...
end
```

I'm not sure there's value in having `let` params, but it does feel a bit odd to have the behavior differ significantly from the way other blocks work. It also means that a `let` block doesn't work at all like the locals at the beginning of a function (which means function inlining doesn't work naturally either).

Reply:
I missed that too. May need to revisit.